samples/psa/persistent_key/README.rst

```diff
@@ -0,0 +1,68 @@
+.. zephyr:code-sample:: persistent_key
+   :name: Persistent key
+
+   Manage and use persistent keys via the Platform Security Architecture (PSA) Crypto API.
+
+Overview
+********
+
+This sample demonstrates how to use the :ref:`PSA Crypto API <psa_crypto>` to generate and use persistent keys.
+
+Requirements
+************
+
+In addition to the PSA Crypto API, an implementation of the PSA Internal Trusted Storage (ITS) API (for storage of the persistent keys) must be present for this sample to work.
+It can be provided by:
+
+* :ref:`tfm` (TF-M), for platforms supporting it.
+* The :ref:`secure storage subsystem <secure_storage>`, for the other platforms.
+
+Building
+********
+
+This sample is located in :zephyr_file:`samples/psa/persistent_key`.
+
+Different configurations are defined in the :file:`sample.yaml` file.
+You can use them to build the sample, depending on the platform to be built for, as follows:
+
+.. tabs::
+
+   .. tab:: TF-M
+
+     For platforms with TF-M:
+
+      .. zephyr-app-commands::
+         :zephyr-app: samples/psa/persistent_key
+         :tool: west
+         :goals: build
+         :board: <ns_platform>
+         :west-args: -T sample.psa.persistent_key.tfm
```

Comment:
Is using of -T mandatory?
What happens if to build it without the test item argument?

Reply:
It's not mandatory per se, it's just a convenient way to pick up the configuration. If you don't specify it you'll need to manually feed the needed Kconfig fragments as listed in the `sample.yaml` file depending on your build.

Reply:
Samples should be buildable & functional without -T or other manual feeding.
Can it be a default sample configuration?
It is looks a bit strange that samples are using -T (test item). 
Probably you need to separate samples and tests.

Reply:
> Samples should be buildable & functional without -T or other manual feeding.

Is this a rule stated somewhere?

If the samples in this PR were to be buildable without configuration argument, there would need to be duplicated `.conf` files for every supported platform.
This would require explicitly listing the supported platforms, and duplicating the different possible configurations, which to me sounds really not good.
The `sample.yaml` file allows to nicely and automatically filter and feed the proper configuration.

The samples are default-buildable for NS platforms, but not for others because they require some more Kconfig options.

Reply:
> It looks a bit strange that samples are using -T (test item).
Probably you need to separate samples and tests.

`-T` in `west build` doesn't mean that it's a test, it just means (from `west build --help`): `Build based on test data in testcase.yaml or sample.yaml.`

Reply:
> > Samples should be buildable & functional without -T or other manual feeding.
> 
> Is this a rule stated somewhere?
> 
> If the samples in this PR were to be buildable without configuration argument, there would need to be duplicated `.conf` files for every supported platform. This would require explicitly listing the supported platforms, and duplicating the different possible configurations, which to me sounds really not good. The `sample.yaml` file allows to nicely and automatically filter and feed the proper configuration.
> 
> The samples are default-buildable for NS platforms, but not for others because they require some more Kconfig options.

- --test-item is used as a build configuration. 
- the samples are tests.

It's better to wait for others' feedback.

Reply:
> `-T` in `west build` doesn't mean that it's a test,

OK '-T' is named as "test-item" and points to a test data ;)

Reply:
> - --test-item is used as a build configuration.
> - the samples are tests.

You're right.
I am planning on adding actual tests, but they will be more unit tests to really push and test the functionality.
Those samples, even though they have a test aspect to them, demonstrate usage. So they do serve the purpose of a sample.
And I think that other samples elsewhere in Zephyr are also used as tests.

Reply:
Stumbled upon this: https://docs.zephyrproject.org/latest/samples/sample_definition_and_criteria.html#sample-criteria
To me the samples here do fulfill all the mentioned criteria.

Reply:
> Stumbled upon this: https://docs.zephyrproject.org/latest/samples/sample_definition_and_criteria.html#sample-criteria To me the samples here do fulfill all the mentioned criteria.

Gray zone: "Samples must not be used as a testcase."

Do you still insist that the Zephyr sample **cannot** be functional without --test-item (as a required parameter)?
Do you know about other existing samples with this approach?
Ok to use --test-item for different optional test configurations, but a sample/test should work also with a default prj.conf.

Reply:
> Gray zone: "Samples must not be used as a testcase."

Below this statement is explicitly mentioned:

>Samples must not use Zephyr’s testing framework.
>
>    - Must not use [CONFIG_ZTEST](https://docs.zephyrproject.org/latest/kconfig.html#CONFIG_ZTEST)
>
>    - Must not use zasserts in samples.
>
>If a sample can provide output that can be verified, then output should be evaluated against expected value to have some level of testing for the sample itself.

and a bit lower (in `2. Twister should be able to build every sample`):
> Sample output can be validated leveraging the harness_config regex option, wherever applicable.

<br>

> Do you still insist that the Zephyr sample cannot be functional without --test-item (as a required parameter)?

Not necessarily this parameter, as the alternative is to feed the Kconfig fragments to the build command directly (which is basically what `-T, --test-item` allows to do more conveniently).
But yes, I think it makes sense because it's way more scalable and maintainable when a sample has a few different configurations for all the platforms.

>  Do you know about other existing samples with this approach?

No.

> Ok to use --test-item for different optional test configurations, but a sample/test should work also with a default prj.conf.

Currently samples in this PR compile with the default configuration (no additional Kconfig fragment) on NS (TF-M-enabled) platforms.
It's possible to make them compile on all platforms by moving relevant configuration from Kconfig fragments to the `prj.conf` file, but that will output Kconfig warnings on NS platforms (because of trying to enable incompatible Kconfig options), so I don't think we want that.
The "proper" way would be to define platform-specific configurations in a `boards` directory. For that, you need to copy-paste the appropriate configuration for every platform you want to support.

So how would you go about doing this? What platforms would you define? And what would you then say for the platforms it would be possible to build for, but for which a configuration hasn't been manually defined?

And again:
> > Samples should be buildable & functional without -T or other manual feeding.
>
> Is this a rule stated somewhere?

And why does it seem so important if the sample's documentation clearly indicates how the sample should be built?

Reply:
> This would require explicitly listing the supported platforms, and duplicating the different possible configurations, which to me sounds really not good.

why? Why can't you list a couple of examples of the different "extra overlays" that need to be pulled in based on the flavor of board people are targetting?

Reply:
What do you mean exactly with `examples of the different "extra overlays"`? Do you mean documenting the possible build commands with `EXTRA_CONF_FILE`?